[All hosts](polyfills) New PnP article on how to use polyfills to support IE11 (#2683)
OfficeDev/office-js-docs-pr: docs/develop/use-polyfills-to-support-ie11-in-your-office-add-in.md

---
title: Use polyfills to support Internet Explorer 11 (IE11) in your Office Add-in
description: Expand the reach of your Office Add-in to support the large number of customers still on IE11 by using polyfills. 
ms.date: 05/13/2021
localization_priority: Priority
---
# Use polyfills to support Internet Explorer 11 (IE11) in your Office Add-in

> There are still a large number of customers who use the Internet Explorer 11 (IE11) browser, and to reach them, you'll want your Office Add-in to successfully run on it. However, IE11 does not implement many JavaScript features that are available on modern browsers like Microsoft Edge and Google Chrome. This article shows how to use Polyfills in your Office Add-in to work around missing IE11 functionality.

Reviewer: ```suggestion
There are still a large number of customers who use the Internet Explorer 11 (IE11) browser, and to reach them, you'll want your Office Add-in to successfully run on it. However, IE11 does not implement many JavaScript features that are available on modern browsers like Microsoft Edge and Google Chrome. This article shows how to use polyfills in your Office Add-in to work around missing IE11 functionality.
```

## How a polyfill works

> A polyfill is code that implements a feature on web browsers that do not support the feature. For example, in IE11, the `Number.isNan` function does not exist. If someone had code that needed to call `Number.isNan` they could write a polyfill to ensure their code is compatible with IE11. The following code sample shows how to implement a `Number.isNan` polyfill:

Reviewer: ```suggestion
A polyfill is code that implements a feature on web browsers that do not support the feature. For example, in IE11, the `Number.isNan` function does not exist. Therefore, if you have code that needs to call `Number.isNan` you should write a polyfill to ensure your code is compatible with IE11. The following code sample shows how to implement a `Number.isNan` polyfill:
```

```javascript
if(!Number.isNan) { // If the function is not defined
	// Write a definition for Number.isNan
	Number.isNan = function(x) {
		return typeof x === 'number' && x !== x;
	}
}
```

> In the previous example, the code checks if `Number.isNan` exists. If it doesn’t, then we define an implementation for `Number.isNaN` that is used automatically whenever `Number.isNan` is invoked. If the programmer did not include this polyfill, their code would fail to run and throw an exception on certain outdated JavaScript engines.

Reviewer: ```suggestion
In the previous example, the code checks if `Number.isNan` exists. If it doesn’t, then you should define an implementation for `Number.isNaN` that is used automatically whenever `Number.isNan` is invoked. If you don't include this polyfill, your code will fail to run and throw an exception on certain outdated JavaScript engines.
```
Sorry, should have caught this in my first edit pass! Change to first person to be consistent with the rest of the article.

After you add the **core-js** library to your Office Add-in project, you can add polyfills for all **core-js** features by importing the library at the top of your entry point. The following example shows how to add the **core-js** library to your script. The `run` function shows an example of calling the `startsWith()` method. The **core-js** library contains numerous polyfills, including the `startsWith()` method.

```javascript
import "core-js";
// import "core-js/features/string/starts-with"

Office.onReady((info) => {
  if (info.host === Office.HostType.Excel) {
    document.getElementById("sideload-msg").style.display = "none";
    document.getElementById("app-body").style.display = "flex";
    document.getElementById("run").onclick = run;
  }
});

export async function run() {
  try {
      console.log('planner'.startsWith('plan'));
    } catch (error) {
    console.error(error);
  }
}
```

> The previous example is based on the [Build an Excel task pane add-in](../quickstarts/excel-quickstart-jquery.md) quick start.

Reviewer: Consider moving this to before the example so you don't have two consecutive lines that start with "in the previous example."

In the previous example, the `startsWith()` method works for both current browsers and for IE11. Current browsers don’t need the polyfill provided by **core-js**, while IE11 will use the imported polyfill.
Importing the entire **core-js** library will import all **core-js** features. If you’re concerned about space, then import only the polyfills that your Office Add-in requires. The [core-js documentation](https://github.com/zloirock/core-js#commonjs-api) contains instructions for this. As an example, the previous code shows how to include the required imports `core-js/features/string/starts-with` instead of importing `core-js`.

> The **core-js** library typically takes care of most polyfills that you need. There are a few exceptions detailed in the [Missing Polyfills](https://github.com/zloirock/core-js#missing-polyfills) section of the [core-js documentation](https://github.com/zloirock/core-js#commonjs-api). Notably, it does not support `fetch`, but you can use the [GitHub window.fetch JavaScript polyfill](https://github.com/github/fetch).

Reviewer: ```suggestion
The **core-js** library typically takes care of most polyfills that you need. There are a few exceptions detailed in the [Missing polyfills](https://github.com/zloirock/core-js#missing-polyfills) section of the [core-js documentation](https://github.com/zloirock/core-js#commonjs-api). Notably, it doesn't support `fetch`, but you can use the [GitHub window.fetch JavaScript polyfill](https://github.com/github/fetch).
```

> If you need to use any methods that don’t work on IE11 and are not supported by these or other libraries, you will need to write your own custom polyfill, as detailed previously.

Reviewer: ```suggestion
If you need to use any methods that don’t work on IE11 and are not supported by these or other libraries, you'll need to write your own custom polyfill, as detailed previously.
```
